src/format.ts

```diff
@@ -110,8 +122,127 @@ function checkFormat(srcFiles: string[], baseArgs: string[]): Promise<boolean> {
+    replacements[i] = {
+      offset: Number(offsets[i]),
+      length: Number(lengths[i]),
+      fix: entities.decodeXML(fixes[i])
+    };
+  }
+  return replacements;
+}
+
+/**
+ * Gets an object containing the differences between the original file and after
+ * changes have been made
+ *
+ * @param file
+ * @param replacements array of all the formatting issues within in the file
+ */
+async function getDiffObj(
+    file: string, replacements: replacement[]): Promise<jsdiff.IUniDiff> {
+  const text = await readFilep(file, 'utf8');
+
+  const fixed = performFixes(text, replacements);
+  const diff = jsdiff.structuredPatch(
+      file, 'no new file', text, fixed, 'no old header', 'no new header',
+      {context: 3});
+  jsdiff.applyPatch('diff', diff);
+  return Promise.resolve(diff);
+}
+
+/**
+ * Performs formatting fixes to the original string 🌷
+ *
+ * @param data original string
+ * @param errOffset start index of the formatting issue
+ * @param errLength length of formatting issue
+ * @param replacements string that resolves formatting issue
+ */
+function performFixes(data: string, replacements: replacement[]) {
+  const replaced: string[] = [];
+  replaced.push(substring(data, 0, replacements[0].offset));
+  for (let i = 0; i < replacements.length - 1; i++) {
+    replaced.push(replacements[i].fix);
+    replaced.push(substring(
+        data, replacements[i].offset + replacements[i].length,
+        replacements[i + 1].offset));
+  }
+  replaced.push(replacements[replacements.length - 1].fix);
+  replaced.push(substring(
+      data,
+      replacements[replacements.length - 1].offset +
+          replacements[replacements.length - 1].length,
+      Buffer.byteLength(data, 'utf8')));
+  return replaced.join('');
+}
+
+/**
+ * Prints the lines with formatting issues
+ *
+ * @param file file to format
+ * @param diffs contains all information about the formatting changes
+ * @param options
+ */
+function printDiffs(diffs: jsdiff.IUniDiff, options: Options) {
+  options.logger.log(chalk.inverse.bold(diffs.oldFileName));
+  diffs.hunks.forEach((diff: JsDiff.IHunk) => {
+    options.logger.log(chalk.bold(
+        '  Lines: ' + diff.oldStart + '-' + (diff.oldStart + diff.oldLines)));
+
+    diff.lines.forEach(function(line: any) {
+      if (line[0] === '-') {
+        options.logger.log('   ' + chalk.red(line));
+      } else if (line[0] === '+') {
+        options.logger.log('   ' + chalk.green(line));
+      } else {
+        options.logger.log('   ' + chalk.black(line));
+      }
+    });
+    options.logger.log('\n');
+  });
+}
+
+/**
+ * Substring using bytes
+ *
+ * @param str original string
+ * @param indexStart position where to start extraction
+ * @param indexEnd position (up to, but not including) where to end extraction
+ * @param encoding
+ */
+function substring(
+    str: string, indexStart: number, indexEnd: number, encoding = 'utf8') {
+  return Buffer.from(str, encoding)
```

Comment:
This looks odd. Since we already had a string – we already had decoded it using whatever encoding that was necessary. Why are we converting to a buffer and then converting back to a string? This does extra work in encoding/decoding, and will be buggy in cases where a different encoding ought to have been used (as rare as it might be).

Reply:
There isn't an available `bytesSubstring` function out there. I was thinking that we should publish this as separate module (it does in fact have subtle differences from `substring`).

Given that this function is only used for showing output, and that JS files are _generally_ written with UTF-8 encoding, I feel like it's acceptable to say (for now) that if your file is encoded differently, you're on your own. That being said, any static tool that prints code snippets is going to have to do the same operation somewhere, so maybe we should look to `tslint` or others for inspiration. I wouldn't be too surprised if `tslint` makes the UTF-8 assumption as well.

Reply:
Discovered this: https://github.com/palantir/tslint/issues/2810